src/main/java/org/embulk/util/config/Compat.java

```diff
@@ -113,21 +127,50 @@ static ObjectNode rebuildObjectNode(final DataSource source) throws IOException
-            throw new IllegalStateException("DataSource(Impl)#toJson() is not accessible.", ex);
+            // "org.embulk.util.config.DataSourceImpl" fo embulk-util-config 0.1.4 or earlier was package-private,
+            // then invoking "toJson" of "org.embulk.util.config.DataSourceImpl" caused IllegalAccessException.
+            // https://github.com/embulk/embulk-util-config/pull/20
+            //
+            // Pass-through to the next step to setAccessible(true) for the "toJson" Method instance.
         }
 
-        if (jsonStringObject == null) {
-            throw new NullPointerException("DataSource(Impl)#toJson() returned null.");
-        }
-        if (!(jsonStringObject instanceof String)) {
-            throw new ClassCastException(
-                    "DataSource(Impl)#toJson() returned not a String: "
-                    + jsonStringObject.getClass().getCanonicalName());
-        }
+        logger.warn("DataSource(Impl)#toJson is not accessible unexpectedly. DataSource class: {}, toJson method: {}",
+                    source.getClass(), toJson);
+        logger.warn("The plugin or another plugin is estimated to be using embulk-util-config:0.1.4 or earlier.");
+        logger.warn("To workaround it, trying to invoke toJson forcibly by Method#setAccessible(true).");
 
-        return Optional.of((String) jsonStringObject);
+        synchronized (toJson) {
+            try {
+                toJson.setAccessible(true);
+            } catch (final SecurityException ex) {
+                throw new IllegalStateException(
+                        "Method#setAccessible(true) is forbidden for thie method DataSource(Impl)#toJson. "
+                            + "DataSource class: " + source.getClass().toString()
+                            + "toJson method: " + toJson.toString(),
+                        ex);
+            }
+
+            try {
+                return toJson.invoke(source);
+            } catch (final InvocationTargetException ex) {
+                final Throwable targetException = ex.getTargetException();
+                if (targetException instanceof RuntimeException) {
+                    throw (RuntimeException) targetException;
+                }
+                if (targetException instanceof Error) {
+                    throw (Error) targetException;
+                }
+                throw new IllegalStateException("DataSource#toJson() threw unexpected Exception.", targetException);
+            } catch (final IllegalAccessException ex) {
+                throw new IllegalStateException(
+                        "DataSource(Impl)#toJson is not accessible even after Method#setAccessible(true). "
+                            + "DataSource class: " + source.getClass().toString()
+                            + "toJson method: " + toJson.toString(),
+                        ex);
+            } finally {
+                toJson.setAccessible(false);
+            }
+        }
```

Comment:
Q: If we narrow the scope of `toJson` by obtaining a new local variable `Method toJson = getToJsonMethod(source)`, is there a need for synchronizing and restore the accessibility here? In case there's no need, do you still consider it potentially dangerous?

Reply:
I thought it can still be dangerous because:
* `Method#setAccessible` is modifying the "method".
* The `toJson` instance, which is retrieved by `Class#getMethod("toJson")`, is expected to be the same instance if another `Class#getMethod("toJson")` is called in another thread.

Yeah, we have to consider a lot of possibility with this type of hacks -- then, as I mentioned, I'm not very positive to merge this.